File: csharp/src/GuardianForge.Agents/Evaluation.cs
using GuardianForge.Core;

namespace GuardianForge.Agents;

/// <summary>The governance scorecard: runs every failure scenario with the deterministic
/// mock supervisor and checks detection, response correctness, false positives, and audit
/// integrity.</summary>
public static class Evaluation
{
    public sealed class Report
    {
        public int Scenarios { get; set; }
        public int DetectCorrect { get; set; }
        public int TypeCorrect { get; set; }
        public int FalsePositives { get; set; }
        public bool AuditIntact { get; set; } = true;
        public List<(string Id, bool Intervened, string Type, bool Correct)> Cases { get; } = new();

        public double DetectionAccuracy => Scenarios == 0 ? 0 : (double)DetectCorrect / Scenarios;
    }

    public static async Task<Report> RunAsync()
    {
        var now = () => DateTimeOffset.FromUnixTimeSeconds(0);
        var rep = new Report();

        foreach (var sc in Scenarios.All())
        {
            rep.Scenarios++;
            var fleet = new Fleet();
            var sup = new Supervisor(new MockLlm());
            var eng = new Engine(Scenarios.DefaultPolicies(), fleet, sup, now);

            Intervention? last = null;
            foreach (var ev in sc.Events)
            {
                var outcome = await eng.ProcessAsync(ev);
                if (outcome.Intervention != null) last = outcome.Intervention;
            }

            var intervened = last != null;
            var detectOk = intervened == sc.ExpectIntervene;
            if (detectOk) rep.DetectCorrect++;
            if (intervened && !sc.ExpectIntervene) rep.FalsePositives++;

            var gotType = last?.Type ?? "";
            bool typeOk;
            if (sc.ExpectIntervene)
                typeOk = intervened && gotType == sc.ExpectType;
            else
                typeOk = !intervened;
            if (typeOk) rep.TypeCorrect++;

            if (!eng.Audit.Verify()) rep.AuditIntact = false;
            rep.Cases.Add((sc.Id, intervened, gotType, detectOk && typeOk));
        }
        return rep;
    }

    public static string Format(Report r)
    {
        var typeAcc = r.Scenarios == 0 ? 0 : 100.0 * r.TypeCorrect / r.Scenarios;
        return
            "GuardianForge Governance Evaluation\n\n" +
            $"Scenarios:                 {r.Scenarios}\n" +
            $"Detection accuracy:        {100.0 * r.DetectionAccuracy:0}%\n" +
            $"Intervention correctness:  {typeAcc:0}%\n" +
            $"False positives:           {r.FalsePositives}\n" +
            $"Audit chain intact:        {(r.AuditIntact ? "yes" : "NO")}\n";
    }
}

File: rust/src/evaluation.rs
//! The governance scorecard (pure scoring + formatting), ported from Go `internal/eval`.
//!
//! This is the pure part of the Go `eval` package: the [`Report`] and [`Case`] aggregates,
//! the detection-accuracy ratio, and the stable text [`Report::format`]. The scenario driver
//! (`Run`) is intentionally excluded - it wires the fleet, agents, LLM, and runtime, which
//! remain in Go.

/// Round to the nearest integer, ties to even - matching Go's `%.0f` (for `x >= 0`).
///
/// Note: the existing C# mirror formats with `{:0}`, which rounds half away from zero and
/// therefore diverges on exact-`.5` percentages (e.g. 12.5). Go is the stated reference, so
/// this port rounds half to even.
fn round_half_even(x: f64) -> i64 {
    let floor = x.floor();
    let diff = x - floor;
    let f = floor as i64;
    if diff < 0.5 {
        f
    } else if diff > 0.5 {
        f + 1
    } else if f % 2 == 0 {
        f
    } else {
        f + 1
    }
}

/// One scenario's outcome.
#[derive(Debug, Clone, PartialEq, Default)]
pub struct Case {
    /// The scenario identifier.
    pub id: String,
    /// Whether GuardianForge intervened.
    pub intervened: bool,
    /// The intervention type taken, if any.
    pub r#type: String,
    /// Whether the outcome matched the expectation.
    pub correct: bool,
}

/// The aggregate scorecard.
#[derive(Debug, Clone, PartialEq)]
pub struct Report {
    /// The number of scenarios run.
    pub scenarios: i64,
    /// The number where intervene/observe matched truth.
    pub detect_correct: i64,
    /// The number where the response type was correct.
    pub type_correct: i64,
    /// The number of healthy agents wrongly intervened on.
    pub false_positives: i64,
    /// Whether the audit chain stayed intact.
    pub audit_intact: bool,
    /// The per-scenario outcomes.
    pub cases: Vec<Case>,
}

impl Default for Report {
    fn default() -> Self {
        Report {
            scenarios: 0,
            detect_correct: 0,
            type_correct: 0,
            false_positives: 0,
            audit_intact: true,
            cases: Vec::new(),
        }
    }
}

impl Report {
    /// The fraction of scenarios where intervene/observe matched truth (0 if none).
    pub fn detection_accuracy(&self) -> f64 {
        if self.scenarios == 0 {
            return 0.0;
        }
        self.detect_correct as f64 / self.scenarios as f64
    }

    /// Render the scorecard in a stable, readable form.
    pub fn format(&self) -> String {
        let intact = if self.audit_intact { "yes" } else { "NO" };
        let type_acc = if self.scenarios > 0 {
            100.0 * self.type_correct as f64 / self.scenarios as f64
        } else {
            0.0
        };
        let mut s = String::from("GuardianForge Governance Evaluation\n\n");
        s.push_str(&format!("{:<27}{}\n", "Scenarios:", self.scenarios));
        s.push_str(&format!(
            "{:<27}{}%\n",
            "Detection accuracy:",
            round_half_even(100.0 * self.detection_accuracy())
        ));
        s.push_str(&format!(
            "{:<27}{}%\n",
            "Intervention correctness:",
            round_half_even(type_acc)
        ));
        s.push_str(&format!(
            "{:<27}{}\n",
            "False positives:", self.false_positives
        ));
        s.push_str(&format!("{:<27}{}\n", "Audit chain intact:", intact));
        s
    }
}
